tests/model_selection/test_cross_validate.py

```diff
@@ -0,0 +1,262 @@
+                ],
+            ),
+        ),
+    )
+    def test_happy_path(
+        self, items_to_recommend: tp.Optional[ExternalIds], expected_metrics: tp.List[tp.Dict[str, tp.Any]]
+    ) -> None:
+        splitter = LastNSplitter(n=1, n_splits=2, filter_cold_items=False, filter_already_seen=False)
+
+        actual = cross_validate(
+            dataset=self.dataset,
+            splitter=splitter,
+            metrics=self.metrics,
+            models=self.models,
+            k=2,
+            filter_viewed=False,
+            items_to_recommend=items_to_recommend,
+        )
+
+        expected = {
+            "splits": [
+                {
+                    "i_split": 0,
+                    "test": 2,
+                    "test_items": 2,
+                    "test_users": 2,
+                    "train": 2,
+                    "train_items": 2,
+                    "train_users": 2,
+                },
+                {
+                    "i_split": 1,
+                    "test": 4,
+                    "test_items": 3,
+                    "test_users": 4,
+                    "train": 6,
+                    "train_items": 2,
+                    "train_users": 4,
+                },
+            ],
+            "metrics": expected_metrics,
+        }
+
+        assert actual == expected
+
+    def test_happy_path_with_features(self) -> None:
+        splitter = LastNSplitter(n=1, n_splits=2, filter_cold_items=False, filter_already_seen=False)
+
+        models: tp.Dict[str, ModelBase] = {
+            "als": ImplicitALSWrapperModel(AlternatingLeastSquares(factors=2, iterations=2, random_state=42)),
+        }
+
+        actual = cross_validate(
+            dataset=self.featured_dataset,
+            splitter=splitter,
+            metrics=self.metrics,
+            models=models,
+            k=2,
+            filter_viewed=False,
+        )
+
+        expected = {
+            "splits": [
+                {
+                    "i_split": 0,
+                    "test": 2,
+                    "test_items": 2,
+                    "test_users": 2,
+                    "train": 2,
+                    "train_items": 2,
+                    "train_users": 2,
+                },
+                {
+                    "i_split": 1,
+                    "test": 4,
+                    "test_items": 3,
+                    "test_users": 4,
+                    "train": 6,
+                    "train_items": 2,
+                    "train_users": 4,
+                },
+            ],
+            "metrics": [
+                {"model": "als", "i_split": 0, "precision@2": 0.5, "recall@1": 0.0},
+                {"model": "als", "i_split": 1, "precision@2": 0.375, "recall@1": 0.25},
+            ],
+        }
+
+        assert actual == expected
+
+    def test_fail_with_cold_users(self) -> None:
+        splitter = LastNSplitter(n=1, n_splits=2, filter_cold_users=False)
+
+        with pytest.raises(KeyError):
```

Comment:
I guess it is supposed to be NotImplementedError. And we need to check splitter settings and raise this in the beginning of cross validation. Otherwise it is a very unexpected behaviour.

Reply:
It's supposed to be... Yeah, maybe. But it's not. This error is coming from models and I won't change it now.

What about splitter settings, so, I don't think we should control it here because it's not about cross validation. And it's possible for user to write their own model that support recommendations for cold users

I'll add warning